### PR TITLE
[codegen][fortran] improvements in casting

### LINKED ISSUE #2815: [Fortran] Extra brackets in the binary expression
```fortran
program expr2
implicit none

contains
elemental real(8) function dabs(x) result(r)
    real(8), intent(in) :: x
    if (x > 0) then
        r = x
    else
        r = 0 - x
    end if
end function dabs
end program
```
`--show-fortran`
```fortran
program expr2
implicit none

contains

elemental real(8) function dabs(x) result(r)
    real(8), intent(in) :: x
    if (x > (0)) then ! <==
        r = x
    else
        r = (0) - x ! <== 
    end if
end function dabs
end program expr2
```

### PR DESCRIPTION
**Pull request by gxyd**
Also fixes #2815 

ToDo:
* Add test cases
* Understand the correct precedence value

### CODE REVIEW

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 1]
```diff
@@ -1669,165 +1669,76 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
-                    case 4: r = "real(" + s + ", " + "kind=dest_kind" + ")"; break;
-                    case 8: r = "real(" + s + ", " + "kind=dest_kind" + ")"; break;
-                    default: throw CodeGenError("Cast LogicalToReal: Unsupported Kind " + std::to_string(dest_kind));
+
+            case ASR::cast_kindType::ComplexToComplex:
+            case ASR::cast_kindType::IntegerToComplex:
+                if (dest_kind == 4 || dest_kind == 8) {
+                    fortran_code = "cmplx(" + fortran_code + ", 0.0, kind=" + std::to_string(dest_kind) + ")";
+                } else {
+                    throw CodeGenError("Unsupported Kind " + std::to_string(dest_kind));
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_fortran.cpp:1695]: I need to improve upon the error message here, since there isn't currently any "string" representation of enumerators of `cast_kindType`

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 2]
```diff
@@ -28,7 +28,7 @@ enum Precedence {
 class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
 {
 public:
-    std::string s;
+    std::string fortran_code;
```
Comment (certik) [inline src/libasr/codegen/asr_to_fortran.cpp:31]: Please send a separate PR with just this change.

Comment (gxyd) [reply src/libasr/codegen/asr_to_fortran.cpp:31]: Opened a separate PR here: https://github.com/lfortran/lfortran/pull/3420

File: tests/logical_to_int_cast_fortran_codegen.f90 [hunk 3]
```diff
@@ -0,0 +1,14 @@
+program ImplicitLogicalToIntCastExample
```
Comment (gxyd) [inline tests/logical_to_int_cast_fortran_codegen.f90:1]: I've opened a separate issue for casting from `logical` to `int` #3436

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 4]
```diff
@@ -1672,155 +1672,118 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
-                last_expr_precedence = 2;
-                break;
-            }
-            case (ASR::cast_kindType::ComplexToLogical) : {
-                s = "(bool)(" + s + ")";
-                last_expr_precedence = 2;
+                last_expr_precedence = Precedence::Ext;
                 break;
             }
             case (ASR::cast_kindType::IntegerToCharacter) : {
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_fortran.cpp:1767]: I doubt that even this is allowed in Fortran. Do we delete this as well?

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 5]
```diff
@@ -1672,155 +1672,118 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
+                    s = "char(" + s + ", " + "kind=" + std::to_string(dest_kind) + ")";
+                } else {
+                    throw CodeGenError("Cast IntegerToCharacter: Unsupported Kind " + \
+                            std::to_string(dest_kind));
                 }
-                last_expr_precedence = 2;
+                last_expr_precedence = Precedence::Ext;
                 break;
             }
             case (ASR::cast_kindType::CharacterToInteger) : {
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_fortran.cpp:1778]: I can't think of an example for this either where fortran would allow the casting. I came across this:

https://stackoverflow.com/questions/9900417/character-to-integer-conversion-in-fortran

but it's not "implicit casting" as such

Is this needed?

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 6]
```diff
@@ -1672,155 +1672,118 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
-                last_expr_precedence = 2;
-                break;
-            }
-            case (ASR::cast_kindType::RealToLogical) : {
-                s = "(bool)(" + s + ")";
-                last_expr_precedence = 2;
-                break;
-            }
             case (ASR::cast_kindType::CharacterToLogical) : {
                 s = "(bool)(len(" + s + ") > 0)";
```
[the remark this answers is not in the document]
Comment (gxyd) [reply src/libasr/codegen/asr_to_fortran.cpp:1763]: > Shall we remove this as well?

We probably should. There is not casting allowed for `CharacterToLogical` anyway.

**Comment by gxyd** [comment]
This pull request is ready for review now.

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 7]
```diff
@@ -1667,167 +1667,41 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
+            {ASR::cast_kindType::IntegerToReal, {"real", {1, 2, 4, 8}}},
+            {ASR::cast_kindType::RealToInteger, {"int", {1, 2, 4, 8}}},
+            {ASR::cast_kindType::RealToReal, {"real", {1, 2, 4, 8}}},
+            {ASR::cast_kindType::IntegerToInteger, {"int", {1, 2, 4, 8}}},
+            {ASR::cast_kindType::ComplexToComplex, {"cmplx", {4, 8}}},
+            {ASR::cast_kindType::IntegerToComplex, {"cmplx", {4, 8}}},
+            {ASR::cast_kindType::ComplexToReal, {"real", {4, 8}}},
+            {ASR::cast_kindType::RealToComplex, {"cmplx", {4, 8}}},
+            {ASR::cast_kindType::LogicalToInteger, {"int", {1, 2, 4, 8}}},
+            {ASR::cast_kindType::IntegerToLogical, {"logical", {1, 2, 4, 8}}}
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_fortran.cpp:1685]: I'm not sure how to handle `IntegerToLogical`, currently getting an error:
```bash
[ 74%] Building Fortran object CMakeFiles/logical4.dir/logical4.f90.o
CMakeFiles/logical4.dir/logical4.f90.o.tmp.f90:6:12:

    6 | a = logical(2, kind=4)
      |            1
Error: ‘l’ argument of ‘logical’ intrinsic at (1) must be LOGICAL
CMakeFiles/logical4.dir/logical4.f90.o.tmp.f90:7:12:

    7 | b = logical(-1, kind=4)
      |            1
Error: ‘l’ argument of ‘logical’ intrinsic at (1) must be LOGICAL
CMakeFiles/logical4.dir/logical4.f90.o.tmp.f90:8:12:

    8 | c = logical(0, kind=4)
      |            1
Error: ‘l’ argument of ‘logical’ intrinsic at (1) must be LOGICAL
```

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 8]
```diff
@@ -1667,167 +1667,41 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
-                last_expr_precedence = 2;
-                break;
-            }
-            case (ASR::cast_kindType::LogicalToCharacter) : {
-                s = "char(" + s + ")";
-                last_expr_precedence = 2;
-                break;
-            }
-            case (ASR::cast_kindType::IntegerToLogical) : {
-                // Implicit conversion between integer -> logical
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_fortran.cpp:1772]: I think, previously, we used to pass. Let's do the same for now.

File: src/libasr/codegen/asr_to_fortran.cpp (hunk 7 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_fortran.cpp:1691]:
```suggestion
```
As it is already, let's remove it.

**Review by Thirumalai-Shaktivel — APPROVED** [review]
Otherwise, looks good!

File: src/libasr/codegen/asr_to_fortran.cpp (hunk 8 above)
Comment (gxyd) [reply src/libasr/codegen/asr_to_fortran.cpp:1772]: Ah yes, missed that. Thanks, I've fixed it now.
